File: examples/typescript/order-specification/order/specification.test.ts
import { Order } from "./order";
import { InMemoryOrderRepository } from "./repository";
import {
  andSpecification,
  NotRefundedSpecification,
  PaidSpecification,
  RefundWindowOpenSpecification
} from "./specification";

function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

const now = new Date(Date.UTC(2026, 5, 30, 12, 0, 0));
const repository = new InMemoryOrderRepository([
  Order.create("ord-ts-6001", true, false, new Date(now.getTime() - 24 * 60 * 60 * 1000), new Date(now.getTime() + 48 * 60 * 60 * 1000)),
  Order.create("ord-ts-6002", true, true, new Date(now.getTime() - 24 * 60 * 60 * 1000), new Date(now.getTime() + 48 * 60 * 60 * 1000)),
  Order.create("ord-ts-6003", true, false, new Date(now.getTime() - 96 * 60 * 60 * 1000), new Date(now.getTime() - 2 * 60 * 60 * 1000)),
  Order.create("ord-ts-6004", false, false, new Date(now.getTime() - 24 * 60 * 60 * 1000), new Date(now.getTime() + 48 * 60 * 60 * 1000))
]);

const specification = andSpecification(
  new PaidSpecification(),
  new NotRefundedSpecification(),
  new RefundWindowOpenSpecification(now)
);

const matches = repository.findMatching(specification);
assert(matches.length === 1, `expected 1 refundable order, got ${matches.length}`);
assert(matches[0]?.id() === "ord-ts-6001", `expected ord-ts-6001, got ${matches[0]?.id()}`);

const expiredOrder = Order.create(
  "ord-ts-6005",
  true,
  false,
  new Date(now.getTime() - 96 * 60 * 60 * 1000),
  new Date(now.getTime() - 2 * 60 * 60 * 1000)
);

const refundWindowOpen = new RefundWindowOpenSpecification(now);
assert(!refundWindowOpen.isSatisfiedBy(expiredOrder), "expected expired refund window to fail");

console.log("order-specification tests passed");

File: examples/go/order-specification/order/specification_test.go
package order

import (
	"testing"
	"time"
)

func TestCompositeSpecificationFiltersRefundableOrders(t *testing.T) {
	now := time.Date(2026, time.June, 30, 12, 0, 0, 0, time.UTC)
	repository := NewInMemoryRepository([]Order{
		New("ord-6001", true, false, now.Add(-24*time.Hour), now.Add(48*time.Hour)),
		New("ord-6002", true, true, now.Add(-24*time.Hour), now.Add(48*time.Hour)),
		New("ord-6003", true, false, now.Add(-96*time.Hour), now.Add(-2*time.Hour)),
		New("ord-6004", false, false, now.Add(-24*time.Hour), now.Add(48*time.Hour)),
	})

	specification := And(
		PaidSpecification{},
		NotRefundedSpecification{},
		RefundWindowOpenSpecification{Now: now},
	)

	matches := repository.FindMatching(specification)

	if len(matches) != 1 {
		t.Fatalf("expected 1 refundable order, got %d", len(matches))
	}

	if matches[0].ID() != "ord-6001" {
		t.Fatalf("expected ord-6001, got %s", matches[0].ID())
	}
}

func TestRefundWindowOpenSpecificationRejectsExpiredOrder(t *testing.T) {
	now := time.Date(2026, time.June, 30, 12, 0, 0, 0, time.UTC)
	specification := RefundWindowOpenSpecification{Now: now}
	candidate := New("ord-6003", true, false, now.Add(-96*time.Hour), now.Add(-2*time.Hour))

	if specification.IsSatisfiedBy(candidate) {
		t.Fatal("expected expired refund window to fail")
	}
}
